### PR TITLE
Add integration tests for desire state of world populator

### PR DESCRIPTION
**Pull request by gnufied**
Add integration tests for desire state of world populator
    
This adds tests for code introduced here :
https://github.com/kubernetes/kubernetes/issues/26994
    
Via integration test we can now verify that if pod delete
event is somehow missed by AttachDetach controller - it still
get cleaned up by Desired State of World populator.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37032)

### CODE REVIEW

**Comment by ixdy** [comment]
cc @xiangpengzhao

**Comment by ixdy** [comment]
@k8s-bot ok to test

File: test/integration/volume/attach_detach_test.go
Comment (xiangpengzhao) [inline test/integration/volume/attach_detach_test.go]: nit: lets --> let's

File: test/integration/volume/attach_detach_test.go
Comment (xiangpengzhao) [inline test/integration/volume/attach_detach_test.go]: every 2 minutes

File: test/integration/volume/attach_detach_test.go
Comment (xiangpengzhao) [inline test/integration/volume/attach_detach_test.go]: Where is the `err` from ?

**Comment by xiangpengzhao** [comment]
A few inline comments. @gnufied

**Comment by saad-ali** [comment]
@rkouj Could you review this PR please.

**Comment by gnufied** [comment]
@xiangpengzhao review comments addressed. PTAL

**Review by xiangpengzhao — APPROVED** [review]
LGTM :)

File: test/integration/volume/persistent_volumes_test.go
Comment (rootfs) [inline test/integration/volume/persistent_volumes_test.go]: why rename and relocate?

File: test/integration/volume/attach_detach_test.go
Comment (rootfs) [inline test/integration/volume/attach_detach_test.go]: 2016

**Comment by rootfs** [comment]
lgtm

File: test/integration/volume/attach_detach_test.go
Comment (rkouj) [inline test/integration/volume/attach_detach_test.go]: What is `s` supposed to be ?
Could you please use a descriptive name for the variable ?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go
Comment (rkouj) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: A comment describing the function here (in the interface) would be nice.

File: test/integration/volume/attach_detach_test.go
Comment (rkouj) [inline test/integration/volume/attach_detach_test.go]: Please rename `s` to `server`. Makes it easier to read the code below

**Review by rkouj — CHANGES_REQUESTED** [review]

File: test/integration/volume/attach_detach_test.go
Comment (rkouj) [inline test/integration/volume/attach_detach_test.go]: Curious, how were these numbers arrived at ?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go
Comment (rkouj) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: I'm not fully convinced that you need to add the function `GetDesiredStateOfWorld()` to the interface if all it does is return the desiredStateOfTheWorld held by the struct. It definitely has to be a function of `adc` but does it really have to be in the interface ?
I may be wrong and I'm not aware of what the best practice here is, but I feel there might be a better way of doing this.

Comment (gnufied) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: `adc` is not exposed from the package and hence can't be read from the test directly. :( 
That is why we are left with only option of having this function in the interface.  Overall, though I think it should be okay.

File: test/integration/volume/attach_detach_test.go
Comment (gnufied) [inline test/integration/volume/attach_detach_test.go]: I simply used from other tests that do similar thing. I think overall idea is to keep it high enough so as we are not throttling API requests.

File: test/integration/volume/attach_detach_test.go
Comment (gnufied) [inline test/integration/volume/attach_detach_test.go]: I could.. but then as project convention goes it is usually accepted in go (See other test files). But I can change this, if it is a problem.

File: test/integration/volume/attach_detach_test.go
Comment (gnufied) [inline test/integration/volume/attach_detach_test.go]: fixed

File: test/integration/volume/attach_detach_test.go
Comment (rkouj) [inline test/integration/volume/attach_detach_test.go]: Yeah the existing code isn't the best example to base off new changes I guess, so it would be great if you could rename, makes it easier to understand what the function returns rather than opening the function definition and seeing its return type.
Let's make the new code we add to be of better quality :) 

Thanks for your help!

Comment (gnufied) [inline test/integration/volume/attach_detach_test.go]: fixed.

**Comment by gnufied** [comment]
@rkouj fixed some of the review comments. PTAL. The e2e failure looks like a flake related to e2e test timing out. Since this PR does not add any e2e test I think, it is safe to say that - this PR did not exacerbate the problem.

**Comment by rkouj** [comment]
Code structure lgtm.
I'll leave it to @saad-ali to sign off on the functionality as he would have more context.

**Comment by gnufied** [comment]
@saad-ali can you look at it and see if we can merge?

**Comment by saad-ali** [comment]
@gnufied I see that this test relies on multiple channels and waits, which can result in flaky tests. To ensure that test is not flaky, could you please run it in a tight loop back to back for an hour or so and see if it flakes?

**Comment by gnufied** [comment]
Ran the tests in a tight loop for an hour and more and got no failures.  I used following command to run:

```
make test-integration KUBE_GOFLAGS="-v" KUBE_TEST_ARGS="-run ^TestPodDeletionWithDswp$"
```

I guess the only thing that stood out is, following messages in logs:

```
I1122 16:19:01.498012   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:01.598873   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:01.699886   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:01.800804   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:01.901791   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:02.002820   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:02.103796   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:02.204729   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:02.305731   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:02.406826   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
I1122 16:19:02.507904   21805 reconciler.go:222] UpdateNodeStatuses failed with: The version is too old: v0.0.0-master+$Format:%h$
```

This comes from mismatched version info check when reconciler checks k8s version itself. But it doesn't affect the outcome of the test.

File: test/integration/volume/attach_detach_test.go
Comment (rkouj) [inline test/integration/volume/attach_detach_test.go]: A true check should be `len(podsToAdd) != 1` , right ?

I guess it's pretty unlikely that more than one pod would be created.

**Review by rkouj — COMMENTED** [review]
Couple of nits.
No need to change.

**Comment by rkouj** [comment]
You can use [something like this](https://gist.github.com/saad-ali/12743a9d3d9d6556f20f7b7b21646d7c) to run tests in a loop.

**Comment by gnufied** [comment]
I did ran the tests in a tight loop for an hour, using something like:

```ruby
20.times do
        system('make test-integration KUBE_GOFLAGS="-v" KUBE_TEST_ARGS="-run ^TestPodDeletionWithDswp$"')
        if ($? != 0 )
                puts "Running tests failed"
        end
end
```

Pretty similar to saad's script.

File: test/integration/volume/attach_detach_test.go
Comment (gnufied) [inline test/integration/volume/attach_detach_test.go]: not sure what I was thinking, will fix.

**Comment by gnufied** [comment]
rebase done. @saad-ali @rkouj let me know, if anything else is needed here.

**Comment by gnufied** [comment]
@k8s-bot unit test this. Flaked on https://github.com/kubernetes/kubernetes/issues/37040

**Comment by rkouj** [comment]
@k8s-bot unit test this

**Comment by rkouj** [comment]
@k8s-bot unit test this

**Comment by rkouj** [comment]
@k8s-bot unit test this

**Comment by gnufied** [comment]
Looks like something changed in functions the tests were using. I need to look into it.

**Comment by gnufied** [comment]
@rkouj The failures were because of wrong version of api types being used here and then clientset definitions changed in rebase.

I have updated the code to use `/v1/` API types and they should pass now.

**Comment by childsb** [comment]
@k8s-bot  ok to test

**Comment by gnufied** [comment]
Flaked on https://github.com/kubernetes/kubernetes/issues/37361

**Comment by gnufied** [comment]
@rkouj @jingxu97 nudge.

File: test/test_owners.csv
Comment (jsafrane) [inline test/test_owners.csv]: I don't like these changes

**Comment by jsafrane** [comment]
> Via integration test we can now verify that if pod delete
event is somehow missed by AttachDetach controller - it still
get cleaned up by Desired State of World populator.


Can you add this text as comment in the test? It's not obvious what is this test trying to do.

**Comment by rkouj** [comment]
@gnufied Please rebase. Once that is done @saad-ali can help merge it in.

File: test/test_owners.csv
Comment (gnufied) [inline test/test_owners.csv]: yeah I don't like this either. `make verify` appears to be failing without this. I will see what I can do.

**Comment by gnufied** [comment]
@rkouj @saad-ali rebased. PTAL.

**Comment by jsafrane** [comment]
@gnufied, please squash the commits

**Comment by gnufied** [comment]
Flaked on https://github.com/kubernetes/kubernetes/issues/27680 . @jsafrane commits squashed, also added comment in test function btw.

**Comment by jsafrane** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**
